src/renderer/api/kube-watch-api.ts

```diff
@@ -1,201 +1,335 @@
+
+export interface IKubeWatchMessage<T extends KubeObject = any> {
+  data?: IKubeWatchEvent<KubeJsonApiData>
+  error?: IKubeWatchEvent<KubeJsonApiError>;
+  api?: KubeApi<T>;
+  store?: KubeObjectStore<T>;
+}
+
+export interface IKubeWatchSubscribeStoreOptions {
+  preload?: boolean; // preload store items, default: true
+  waitUntilLoaded?: boolean; // subscribe only after loading all stores, default: true
 }
 
-export interface IKubeWatchRouteEvent {
-  type: "STREAM_END";
-  url: string;
-  status: number;
+export interface IKubeWatchReconnectOptions {
+  reconnectAttempts: number;
+  timeout: number;
 }
 
-export interface IKubeWatchRouteQuery {
-  api: string | string[];
+export interface IKubeWatchLog {
+  message: string | Error;
+  meta?: object;
 }
 
 @autobind()
 export class KubeWatchApi {
-  protected evtSource: EventSource;
-  protected onData = new EventEmitter<[IKubeWatchEvent]>();
-  protected subscribers = observable.map<KubeApi, number>();
-  protected reconnectTimeoutMs = 5000;
-  protected maxReconnectsOnError = 10;
-  protected reconnectAttempts = this.maxReconnectsOnError;
+  private cluster: Cluster;
+  private namespaceStore: NamespaceStore;
+
+  private requestId = 0;
+  private isConnected = false;
+  private reader: ReadableStreamReader<string>;
+  private subscribers = observable.map<KubeApi, number>();
+
+  // events
+  public onMessage = new EventEmitter<[IKubeWatchMessage]>();
+
+  @computed get isActive(): boolean {
+    return this.apis.length > 0;
+  }
+
+  @computed get apis(): string[] {
+    const { cluster, namespaceStore } = this;
+    const activeApis = Array.from(this.subscribers.keys());
+
+    return activeApis.map(api => {
+      if (!cluster.isAllowedResource(api.kind)) {
+        return [];
+      }
+
+      if (api.isNamespaced) {
+        return namespaceStore.getContextNamespaces().map(namespace => api.getWatchUrl(namespace));
+      } else {
+        return api.getWatchUrl();
+      }
+    }).flat();
+  }
 
   constructor() {
-    reaction(() => this.activeApis, () => this.connect(), {
+    this.init();
+  }
+
+  private async init() {
+    const { getHostedCluster } = await import("../../common/cluster-store");
+    const { namespaceStore } = await import("../components/+namespaces/namespace.store");
+
+    await namespaceStore.whenReady;
+
+    this.cluster = getHostedCluster();
+    this.namespaceStore = namespaceStore;
+    this.bindAutoConnect();
+  }
+
+  private bindAutoConnect() {
+    const connect = debounce(() => this.connect(), 1000);
+
+    reaction(() => this.apis, connect, {
       fireImmediately: true,
-      delay: 500,
+      equals: comparer.structural,
     });
-  }
 
-  @computed get activeApis() {
-    return Array.from(this.subscribers.keys());
+    window.addEventListener("online", () => this.connect());
+    window.addEventListener("offline", () => this.disconnect());
+    setInterval(() => this.connectionCheck(), 60000 * 5); // every 5m
```

Comment:
If app is minimized you're suggesting to what exactly to do? How it should be related to connection?

Reply:
I think @aleksfront has a point here.. we could optimize this so that Lens stops watching if window is unfocused (or if frame is not active). Anyway it's out of scope for this PR.